src/openms/include/OpenMS/KERNEL/MSExperiment.h

```diff
@@ -392,6 +392,395 @@ namespace OpenMS
+        result[i].resize(stop - start);
+        
+        for (size_t j = start; j < stop; ++j) 
+        {
+            spec_idx_to_range_idx[j].push_back(i);
+        }
+    }
+
+   #pragma omp parallel for schedule(dynamic)
+   for (Int64 i = 0; i < (Int64)spec_idx_to_range_idx.size(); ++i) // OpenMP on windows still requires signed loop variable
+   {
+      const auto& spec = spectra_view[i].get();
+      MSSpectrum::ConstIterator spec_begin = spec.cbegin();
+      MSSpectrum::ConstIterator spec_end = spec.cend();
+
+      for (size_t range_idx : spec_idx_to_range_idx[i]) {
+        const auto& mz_range = mz_rt_ranges[range_idx].first;
+        
+        // Find data points within MZ range
+        auto start_it = spec.PosBegin(spec_begin, mz_range.getMinMZ(), spec_end);
+        auto end_it = start_it;
+        
+        while (end_it != spec_end && end_it->getPosition() < mz_range.getMaxMZ()) 
+        {
+          ++end_it;
+        }
+
+        // Calculate result using provided reduction function
+        result[range_idx][i - rt_ranges_idcs[range_idx].first] = 
+            func_mz_reduction(start_it, end_it);
+      }
+    }
+    return result;
+  }
+
+// Overload without func_mz_reduction parameter (default to SumIntensityReduction). Needed because of template deduction issues
+std::vector<std::vector<MSExperiment::CoordinateType>> aggregate(
+    std::vector<std::pair<RangeMZ, RangeRT>>& mz_rt_ranges,
+    unsigned int ms_level) const
+{
+  return aggregate(mz_rt_ranges, ms_level, SumIntensityReduction());
+}
+
+/**
+ * @brief Extracts extracted ion chromatograms (XICs) from the MSExperiment.
+ *
+ * This function takes a vector of mz_rt_ranges, an ms_level, and a MzReductionFunctionType
+ * and extracts the XICs from the MSExperiment based on the given parameters.
+ *
+ * @param mz_rt_ranges A vector of pairs of RangeMZ and RangeRT representing the m/z and retention time ranges.
+ * @param ms_level The MS level of the spectra to consider.
+ * @param func_mz_reduction The MzReductionFunctionType used to reduce the m/z values.
+ *
+ * @return A vector of MSChromatogram objects representing the extracted XICs.
+ */
+template<class MzReductionFunctionType>
+std::vector<MSChromatogram> extractXICs(
+    std::vector<std::pair<RangeMZ, RangeRT>>& mz_rt_ranges,
+    unsigned int ms_level,
+    MzReductionFunctionType func_mz_reduction) const
+{
+    // Early exit if there are no ranges
+    if (mz_rt_ranges.empty()) 
+    {
+      // likely an error, but we return an empty vector instead of throwing an exception for now
+      return {};
+    }
+
+    // Sort mz_rt_ranges by ascending MinMZ and descending MaxMZ
+    std::sort(mz_rt_ranges.begin(), mz_rt_ranges.end(), 
+        [](const auto& a, const auto& b) 
+        {
+          if (a.first.getMinMZ() != b.first.getMinMZ()) 
+          {
+            return a.first.getMinMZ() < b.first.getMinMZ();
+          }
+          return -a.first.getMaxMZ() < -b.first.getMaxMZ();
```

Comment:
Whut? Is this faster than just using >?
It is definitely less readable.

Reply:
can undo... it became a bit more common with std::tie and a single operator< but I can change

Reply:
Ah I see.

Reply:
I missed that completely.
I would guess this is even slower than `>`.